[ci] Use auxclasspath with regression-tester

adangel opened the pull request:
## Describe the PR

* For testing, this will use the pmd-regression-tester from PR-branch pmd/pmd-regression-tester#72 . If it works, we can merge that regression-tester-pr and point to the master branch
* ~This additionally - a bit unrelated - changes, how we download openjdk11. We now use a fixed URL, so that we don't need to update the version in the script anymore.~ I've cherry-picked these already on master
* Finally, updates the `project-list.xml` configuration to include build-command and auxclasspath-command for both checkstyle and spring-framework. This enables auxclasspath usage. We hopefully see the different between no-typeresolution (current master baseline) and typeresolution (this PR).

Note: the project-list has a minimal set of exclude-patterns (for checkstyle). For PMD7, there are more needed (refs #2808). The complete set is documented in pmd/pmd-regression-tester#72

## Related issues

- Depends on pmd/pmd-regression-tester#48

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [ ] Added (in-code) documentation (if needed)

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This changeset introduces 0 ne...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/3f57327e85bb4ad080faa63983a8c87c/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented:
Finally the travis build went through. I'm not sure, whether it really works reliable. It stopped getting output when analyzing spring-framework. It was supposed to log every two minutes a message ("Still building ....") to actually avoid this exact problem.

Highlights from checkstyle:
* :heavy_check_mark: It seems, that MissingOverride benefits massively from typeresolution (many fixed false-negatives). 
* :x: UselessOverridingMethod seems to have now a new [false negative](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A66) Although I'm not entirely sure about this case: the class is package-private... does that make a difference?
* :heavy_check_mark: There are also fixed examples for UselessOverridingMethod: [finalize is now elevated to public](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A247)
* :x: As well as CompareObjectsWithEquals: [new false negative](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A111)
* :heavy_check_mark: SwitchStatementShouldHaveDefault has a [fixed false-positive](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A200) with enums
* :heavy_check_mark: UnnecessaryFullyQualifiedNames also has fixed false positives, [for example](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A210)
* :x: somehow we lost the ability to detect StringBuffer/StringBuilder, e.g. [InsufficientStringBufferDeclaration](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A228) and [ConsecutiveStringLiterals](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A229)
* :heavy_check_mark: Fixed false negative with [AvoidUncheckedExceptionsInSignatures](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A317)
* :x: a false negative now with [UseTryWithResources](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A360)

Highlights from spring-framework:
* :x: there seem to be really problems for StringBuilder/StringBuffer: [AppendCharacterWithChar](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/spring-framework/index.html#A1) - that's definitively now a false negative
* :heavy_check_mark: We have some fixed false negatives in [AvoidInstantiationInLoops](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/spring-framework/index.html#A20) - although I don't know why...
* :heavy_check_mark: Many fixed [UnusedImports](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/spring-framework/index.html#A68) since we have now junit on the auxclasspath
* :x: new false positive [JUnitAssertionsShouldIncludeMessage](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/spring-framework/index.html#A6120)
* :warning: :x: There are processing errors, e.g. [TypeNotPresentException](https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/spring-framework/index.html#B1). That's the reason, why there are missing violations for these files (e.g. EhCacheManagerFactoryBean.java). This exception is very similar to a LinkageError and we should handle it probably the same way.
    * The dependency to "net.sf.ehcache:ehcache:2.10.4" is actually optional and it seems, that optional dependencies are not considered by the "createSquishClasspath" task I added... I'll add those...

adangel commented:
A very inaccurate estimate of the build times:
* Without regression testing (e.g. https://travis-ci.com/github/pmd/pmd/builds/194443832): 14min
* regression testing without typeres (e.g. https://travis-ci.com/github/pmd/pmd/builds/193466250): 23min
* regression testing with typeres (incl. building checkstyle+spring) (e.g https://travis-ci.com/github/pmd/pmd/builds/194362457): 31min
